ffi/ffi.cpp

```diff
@@ -0,0 +1,256 @@
+#include "ffi.h"
+#include "ffi_serialization.hpp"
+#include <libff/algebra/curves/bls12_377/bls12_377_pp.hpp>
+#include <libff/algebra/curves/bw6_761/bw6_761_pp.hpp>
+
+namespace libff
+{
+
+namespace ffi
+{
+
+// Generic functions to be used by the entry points
+
+template<typename ppT>
+bool g1_add(
+    const void *a_g1,
+    size_t a_g1_size,
+    const void *b_g1,
+    size_t b_g1_size,
+    void *out_g1,
+    size_t out_g1_size)
+{
+    libff::G1<ppT> a;
+    libff::G1<ppT> b;
+    if (group_element_read(a, a_g1, a_g1_size) &&
+        group_element_read(b, b_g1, b_g1_size)) {
```

Comment:
For these long if conditions I think it's worth breaking after the closing parenthesis of the `if` statement condition to enhance readability as we are not very helped by the indentation. (Else, extra variables to hold the boolean result of the if statement condition can be useful, though not strictly necessary if the block structure is clear)

Reply:
Yes, I think ffi places the braces on the next line pretty consistently.  Switched to that everywhere.  Thanks.